feat(rpc): Add fields to `getblockchaininfo` RPC output

arya2 opened the pull request:
## Motivation

This PR adds fields that are present on the zcashd `getblockchaininfo` RPC output which are likely required from Zebra's `getblockchaininfo` RPC too (except `chain_work` which should be added as part of https://github.com/ZcashFoundation/zebra/issues/7109).

Closes https://github.com/ZcashFoundation/zebra/issues/8449

## Solution

- Adds the missing fields to the response struct (except `softforks`)
- Adds a state request for querying the disk space usage
- Renames `ValuePoolBalance` to `Balance` and re-uses it for `chain_supply` by summing the amounts of all pools

Related:
- Runs `getblockchaininfo` state queries in parallel,
- Removes `BlockHeader` query, reads tip block time from latest chain tip channel instead, and
- Avoids returning an error when the state is empty (returns genesis/default values instead).

### Tests

- Updated snapshots
  - Ignores the `verificationprogress` field as it depends on the current time
- Updated existing prop tests
- Used shared logic covered by existing tests for the `difficulty` field

### Follow-up Work

- https://github.com/ZcashFoundation/zebra/issues/7109

### PR Author's Checklist

<!-- If you are the author of the PR, check the boxes below before making the PR
ready for review. -->

- [ ] The PR name will make sense to users.
- [ ] The PR provides a CHANGELOG summary.
- [ ] The solution is tested.
- [ ] The documentation is up to date.
- [ ] The PR has a priority label.

### PR Reviewer's Checklist

<!-- If you are a reviewer of the PR, check the boxes below before approving it. -->

- [ ] The PR Author's checklist is complete.
- [ ] The PR resolves the issue.

arya2 committed:
Adds some of the required fields on `getblockchaininfo` output.

arya2 committed:
Adds state request/response variants for querying disk usage

arya2 committed:
Adds `size_on_disk`, `chain_supply`, and `monitored` fields.

arya2 committed:
Updates snapshots

arya2 committed:
fixes prop tests

arya2 committed:
fixes doc lints

arya2 committed:
Adds missing `size()` method

oxarbitrage commented on zebra-rpc/src/methods/tests/snapshot.rs:
```diff
@@ -613,6 +613,12 @@ fn snapshot_rpc_getblockchaininfo(
             }),
+            ".verificationprogress" => dynamic_redaction(|value, _path| {
+                // assert that the value looks like a valid verification progress here
+                assert!(value.as_f64().unwrap() < 1.0);
```
```suggestion
                assert!(value.as_f64().unwrap() <= 1.0);
```

Q: If the chain is in sync, can this value be actually 1.0?

arya2 replied:
Yep, applied as part of https://github.com/ZcashFoundation/zebra/pull/9215/commits/c62b61ef4e25ed4043db5a4ed82dc00963b0de74.

oxarbitrage commented on zebra-state/src/service/finalized_state/disk_db.rs:
```diff
@@ -570,6 +570,27 @@ impl DiskDb {
         );
     }
 
+    /// Returns the estimated total disk space usage of the database.
+    pub fn size(&self) -> u64 {
+        let db: &Arc<DB> = &self.db;
+        let db_options = DiskDb::options();
+        let mut total_size_on_disk = 0;
+        for cf_descriptor in DiskDb::construct_column_families(&db_options, db.path(), &[]).iter() {
+            let cf_name = &cf_descriptor.name();
+            let cf_handle = db
+                .cf_handle(cf_name)
+                .expect("Column family handle must exist");
+
+            total_size_on_disk += db
+                .property_int_value_cf(cf_handle, "rocksdb.total-sst-files-size")
+                .ok()
+                .flatten()
+                .unwrap_or(0);
+        }
+
+        total_size_on_disk
+    }
```
Q: Can we reuse this function [here](https://github.com/ZcashFoundation/zebra/blob/main/zebra-state/src/service/finalized_state/disk_db.rs#L518)?

arya2 replied:
They have less in common than it would seem, using `size()` in `print_db_metrics()` would save one line and I think the current code is simpler.

oxarbitrage commented on zebra-state/src/service/read/difficulty.rs:
```diff
@@ -82,6 +82,7 @@ pub fn get_block_template_chain_info(
 ///
 /// Returns the solution rate per second for the current best chain, or `None` if
 /// the `start_hash` and at least 1 block below it are not found in the chain.
+#[allow(unused)]
```
```suggestion
```

I am not sure if this is needed for some build type, i commented out and built with `cargo build` and there were no complains.

arya2 replied:
It only complains when building with `--no-default-features`.

oxarbitrage reviewed (dismissed):
This looks really good to me. We want to do some manual diff tests after https://github.com/ZcashFoundation/zebra/pull/9216 gets merged in before merging. I left some minor comments/questions but no blockers.

arya2 committed:
Fixes lwd integration test issue by updating get_blockchain_info to fallback on default values instead of returning an error if the state is empty.

Related: Runs state queries in parallel from getblockchaininfo RPC and removes the BlockHeader query by getting the tip block time from the latest chain tip channel.

arya2 committed:
Updates failing proptests

arya2 committed:
fixes lint

arya2 commented:
@Mergifyio refresh
